Author: Sharon Gratch <sgratch@redhat.com>

sgratch opened the pull request:
Add an option to send a Non-Maskable interrupt to a guest Os.

    https://bugzilla.redhat.com/show_bug.cgi?id=1297037

mareklibra commented on pkg/machines/libvirt.es6:
I think backticks are missing here

mareklibra commented on pkg/machines/hostvmslist.jsx:
Please handle case, the provider is not supporting this canSendNMI function, i.e.:

    const sendNMI = config.provider.canSendNMI && config.provider.canSendNMI(state) ? ...

mareklibra commented on pkg/machines/hostvmslist.jsx:
By doing this, the NMI is sent once the button is clicked.

Does Yes/No confirmation make sense?
Since consequences of this function might not be obvious, the user should get proper guiding.

sgratch commented:
#### Since which libvirt version is he 'inject-nmi' virsh command supported? 
 #### (I've already faced an issue on 2-years old Libvirt which needs to be still somehow reasonably handled by our code).

Good question, I checked and it seems supported since livirt 0.9.2 (2011).
https://www.redhat.com/archives/libvirt-users/2011-June/msg00009.html
So I guess we don't have a to worry about that...

#### How difficult is to implement integration test for this? See test/verify/check-machines

Not seems so difficult but I didn't manage to run the test.
the ./test/vm-download failed with:
https://fedorapeople.org/groups/cockpit/images/fedora-24-272dd19c91336c947ac5f3860cc98c064cd8ab54.qcow2.xz

curl: (22) The requested URL returned error: 404 Not Found

#### Please change the commit message.
#### First line should be like:
#### machines: Send Non-Maskable interrupt feature added
#### with more detailed description why adding it and what is it for.

Done.

sgratch commented on pkg/machines/libvirt.es6:
which backticks are missing? I checked that and it seems working.

sgratch commented on pkg/machines/hostvmslist.jsx:
Done

sgratch commented on pkg/machines/hostvmslist.jsx:
Why Yes/No confirmation dialog is not required also for Restart/Shut Down actions?
Do you know of a reference in cockpit code of how it should look like/implemented?

mareklibra commented on pkg/machines/libvirt.es6:
I can't see former code, github shows me that as outdated. 
Anyway, this line seems ok now.

mareklibra commented on pkg/machines/hostvmslist.jsx:
The only reference I can find is use of modal dialog  but I'm not sure this is a pattern.
Since this has wider impact, I'll open discussion with UX designer about something more general.
So let's keep the confirmation for a follow-up.

sgratch commented:
@mareklibra,  I rebased my add_nmi_support branch and added a new test for the new SENDNMI_VM operation.
All your other comments were also verified & done (except the yes/no verification dialog..).

Can you see my changes now?

sgratch commented:
@mareklibra, https://github.com/sgratch/cockpit/pull/1/commits/2dc91523cba5c880d613c51a33c3f2a9a36f0736

mareklibra commented:
@sgratch , to properly rebase your changes on top of last upstream/master development, please try:
- verify your .git/config file contains something like:

      [remote "upstream"]
           url = https://github.com/cockpit-project/cockpit.git
           fetch = +refs/heads/*:refs/remotes/upstream/*

- git checkout add_nmi_support           # your local branch
- git checkout -b upstreamReset          # sort of "save a copy"
- git log                                                 # and note your commit ID, recently 2dc9152
- git checkout add_nmi_support           # go back
- git fetch --all && git reset --hard upstream/master  # update your local branch to upstream/master
- git cherry-pick [YOUR-COMMIT-ID]  # apply your changes; optionally resolve conflicts
- git push --force                                   # update this pull request


With this, you shall end with just a single commit within this PR, not the 250.

sgratch commented:
@mareklibra  The expected behavior after the NMI is sent is not certain and depends on parameters not known to cockpit. Please remember that since we implemented it over libvirt then this feature is very basic and limited. - i.e:
1. Since libvirt doesn't aware if the vm is responsive or not - then nothing to test here since we'll send NMI anyway (if canReset(vmState) returns true).
2. If the guest OS is not installed then sending NMI won't change anything (no OS to handle the NMI). 
3. If the guest OS is installed and NMI is sent then it depends on how the guest OS is configured:
a. if the kernel properties are set to handle NMI (kernel.panic_on_unrecovered_nmi = 1,  
    kernel.unknown_nmi_panic = 1) then the kernel will get into panic mode in case of sending NMI.
b. If the kernel properties are not set then nothing will be changed in VM status.
c. If kernel properties are set to handle NMI and also kdump is enabled then a vmcore file will be created   
     during the crash.
So the vm may stop running and may not - the result is unknown.

That is why I only tested the ability to click the button and that it's supported by the provider and nothing more.

mareklibra commented:
@sgratch , thanks for explanation.

With recent test, we practically just check whether the button is rendered and UI doesn't fail on click, but not the integrated behavior. With respect to what you wrote above, this might make sense. From my point of view, this is good enough.

Anyway if you want, some hacking might bring improvement - the test machine started within check-machines is already given. Considering it's very configuration, does the NMI changes anything? If not, let's keep the test as it is. Otherwise and when not-to-much-complicated, let's check the result.

sgratch commented:
@mareklibra 
**Regarding the integration test for sending NMI feature:**
 I'm not sure how to implement it. You are right that currently only clicking the UI button is checked and we can assume that the tested VM may not respond to the NMI and nothing will be changed in this tested VM state - so nothing to so.

But the problem is that I log "SENDNMI_VM not implemented for this PROVIDER" which is a bit ugly. 
Should I omit this log message and actually do nothing in the provider of cockpitMachinesTestExternalProvider ? 
Or maybe call PROVIDER.actions.updateOrAddVm(..) without changing the state of the vm?
Not sure how to handle this test (It was commented as a problem also in the upstream PR).

**Regarding the 250 commits added to this PR, instead of only my one commit:**
Apparently the reason for that  was that this PR was created as a difference from my master branch (and not the upstream master branch). For verifying that,  I created a new PR in the upstream branch, as you could see...:) 
Do you think I should add a reference to this PR from the new upstream one? 

Thanks!

mareklibra commented:
@sgratch , I have no problem referencing this PR in the upstream PR if you think it can provide additional information.
Anyway, I'm further commenting on that there.
